Adding the three Variable functionality to the Describe> Two\Three Summary dialogue

Ivanluv opened the pull request:
This replaces PR #7465

@lilyclements I have added the Message to display when 2/more categorical variables are selected in two different receivers i.e for the Two and Three Variable. Could you have a look?

lilyclements commented:
@Ivanluv this is working really great.
On the "three variable" option, I don't get the warning message if I put the same variable in the second receiver that is in the first. However, I do get the warning when reopening the dialog. Is it possible to change this?

Ivanluv commented:
> However, I do get the warning when reopening the dialog. Is it possible to change this?

Yes, I have fixed this .The warning to the three variable second receiver is also fixed

rdstern commented:
@Ivanluv this is really getting there.  
a) very minor is that the second categorical is not vertically aligned with the first.
b) when you tick to have percentages, Ok isn't enabled unless you include a factor.  It is possible to have overall percentages, so Ok should be enabled even if the receiver is empty.
c) Check with @lilyclements first, but I am suggesting (even in the general case) that the dialogue only allow a maximum of one factor as the base for the percentages.  If she agrees, then the 3-variable could become a single receiver (as it is in the 2-variable case) 
d) In both the 2-variable and 3-variable cases, please add the `Format Table` button (from the general Summary dialogue).

Ivanluv commented:
> c) Check with @lilyclements first, but I am suggesting (even in the general case) that the dialogue only allow a maximum of one factor as the base for the percentages. If she agrees, then the 3-variable could become a single receiver (as it is in the 2-variable case)

@lilyclements what is your take on this?

lilyclements commented:
> @lilyclements what is your take on this?

@Ivanluv this sounds good to me

Ivanluv commented:
@lilyclements  ,this is how it gets added in the summary table dialogue ie

```
# Code generated by the dialog, Summary\Frequency Tables
frequency_table <- data_book$summary_table(data_name="survey", factors=c("village","variety"), store_results=FALSE, treat_columns_as_factor=FALSE, summaries=count_label)


last_table <- (mmtable2::mmtable(data=frequency_table, cells=value) + mmtable2::header_top_left(variable=village) + mmtable2::header_left_top(variable=variety)) %>% gt::tab_header(title="SUMMARY", subtitle="VILLAGE") %>% gt::tab_style(style=gt::cell_text(font="NULL", size=gt::px(18)), location=gt::cells_title(groups="title"))
data_book$add_table(table_name="last_table", table=last_table, data_name="survey")
data_book$get_tables(data_name="survey", table_name="last_table")


```

how can the we add the format functions in the mapping function below?

```
# Code generated by the dialog, Describe Two Variables
survey <- data_book$get_data_frame(data_name="survey")
multiple_receiver_cols <- c("size")
list_of_tables  <- survey %>% purrr::map(.x=multiple_receiver_cols, .f= ~ data_book$summary_table(data_name="survey", summaries=c("summary_count"), factors=c("variety"), columns_to_summarise=.x)) %>% purrr::map(.x=., .f= ~ mmtable2::mmtable(data=.x, cells=value) + mmtable2::header_top_left(variable="variety") + mmtable2::header_left_top(variable='summary-variable'))



```

rdstern reviewed (changes requested):
@Ivanluv just some minor cosmetic changes and then it can be merged.  Great if @lloyddewit could check the code in parallel.
a) In 2-way frequency tables make the same changes as in the general tables, so they are consistent.  
![image](https://user-images.githubusercontent.com/11226469/181926302-44d3deda-2f4d-4c61-98dd-6d00c4cb12dc.png)
 So on the left, the categorical are not quite aligned
On the right, change `Factors as Percentage `to `Of Factor (Optional) `

b) And make that label the same on 3-way.  And the spacing can be changed for the decimal on 3-way.
c) On the 3-way I think both the second and 3rd variables must currently be factors.  So the only choice is on the first variable.  You fix the second variable currently, but not the third.  For now fix both to be factors.
d) So, on the left, you can move the 2nd and 3rd elements to the lower line.  They ate both categorical and the only choice (currently) is the multiple receiver, which can  (as now, be numeric or categorical.

lloyddewit commented on instat/dlgDescribeTwoVariable.vb:
```diff
@@ -130,464 +181,768 @@ Public Class dlgDescribeTwoVariable
+        grpOptions.Visible = False
+
+        If rdoTwoVariable.Checked Then
+            Select Case strFirstVariablesType
+                Case "numeric"
+                    ucrChkOmitMissing.Visible = True
+                    If strSecondVariableType = "categorical" Then
+                        grpOptions.Visible = True
+                    End If
+            End Select
```
shorten? e.g.
```suggestion
            ucrChkOmitMissing.Visible = strFirstVariablesType = "numeric"
            grpOptions.Visible = IsNumericByCategorical()
```

Ivanluv replied:
Done

lloyddewit commented on instat/dlgDescribeTwoVariable.vb:
```diff
@@ -130,464 +181,768 @@ Public Class dlgDescribeTwoVariable
+                    End If
+            End Select
+        ElseIf rdoThreeVariable.Checked Then
+            If (strFirstVariablesType = "categorical" AndAlso
+            strSecondVariableType = "numeric") OrElse
+            (strFirstVariablesType = "numeric" And
+            strSecondVariableType = "categorical") Then
+                grpOptions.Visible = True
+            Else
+                grpOptions.Visible = False
+            End If
```
shorten? (see example above)

Ivanluv replied:
Done

lloyddewit commented on instat/dlgDescribeTwoVariable.vb:
```diff
@@ -130,464 +181,768 @@ Public Class dlgDescribeTwoVariable
+        If rdoSkim.Checked Then
+            Exit Sub
+        End If
+
+        If IsNumericByCategorical() OrElse
+       IsCategoricalByCategorical() OrElse (IsCategoricalByNumeric() AndAlso
+       rdoThreeVariable.Checked) Then
+            If Not IsNothing(clsJoiningPipeOperator.GetParameter("pipe")) Then
+                ucrBase.clsRsyntax.AddToAfterCodes(clsFormatTableMappingPipeOperator, 2)
             End If
-        Else
-            grpOptions.Visible = False
-            grpSummaries.Visible = False
         End If
```
more readable?
```
        If rdoSkim.Checked OrElse IsNothing(clsJoiningPipeOperator.GetParameter("pipe")) Then
            Exit Sub
        End If

        If IsNumericByCategorical() OrElse IsCategoricalByCategorical() _
                OrElse (IsCategoricalByNumeric() AndAlso rdoThreeVariable.Checked) Then
            ucrBase.clsRsyntax.AddToAfterCodes(clsFormatTableMappingPipeOperator, 2)
        End If
```

Ivanluv replied:
True. Done

lloyddewit commented on instat/dlgDescribeTwoVariable.vb:
```diff
@@ -130,464 +181,768 @@ Public Class dlgDescribeTwoVariable
-            clsCombineMultipleColumnsFunction.ClearParameters()
-            clsCombineFactorsFunction.ClearParameters()
+    Private Sub UpdateCombineFactorFunctions()
+        Dim iPosition As Integer = 0
+        clsCombineMultipleColumnsFunction.ClearParameters()
+        If rdoThreeVariable.Checked OrElse rdoTwoVariable.Checked Then
+            If Not ucrReceiverFirstVars.IsEmpty Then
+                For Each strColumn In ucrReceiverFirstVars.GetVariableNamesList()
+                    clsCombineMultipleColumnsFunction.AddParameter(strColumn, strColumn, bIncludeArgumentName:=False, iPosition:=iPosition)
+                    iPosition += 1
+                Next
+            End If
+        End If
```
shorten?
```suggestion
        If Not rdoThreeVariable.Checked AndAlso (Not rdoTwoVariable.Checked OrElse ucrReceiverFirstVars.IsEmpty) Then
            Exit Sub
        End If
        Dim iPosition As Integer = 0
        For Each strColumn In ucrReceiverFirstVars.GetVariableNamesList()
            clsCombineMultipleColumnsFunction.AddParameter(strColumn, strColumn, bIncludeArgumentName:=False, iPosition:=iPosition)
            iPosition += 1
        Next
```

Ivanluv replied:
Done

lloyddewit commented on instat/dlgDescribeTwoVariable.vb:
```diff
@@ -130,464 +181,768 @@ Public Class dlgDescribeTwoVariable
+    End Sub
+
+    Private Sub HideFormatTableButton()
+        cmdFormatTable.Visible = IsNumericByCategorical() OrElse
+        IsCategoricalByCategorical() OrElse (IsCategoricalByNumeric() AndAlso
+        rdoThreeVariable.Checked)
```
add some indents?
```suggestion
        cmdFormatTable.Visible = IsNumericByCategorical() _
                OrElse IsCategoricalByCategorical() OrElse (IsCategoricalByNumeric() _
                AndAlso rdoThreeVariable.Checked)
```

Ivanluv replied:
Done

lloyddewit commented on instat/dlgDescribeTwoVariable.vb:
```diff
@@ -130,464 +181,768 @@ Public Class dlgDescribeTwoVariable
-        If rdoCustomize.Checked Then
-            If strFirstVariablesType = "categorical" AndAlso strSecondVariableType = "categorical" Then
-                grpDisplay.Visible = True
+        grpThreeVariablePercentages.Visible = False
+        grpTwoVariablePercentages.Visible = False
+        grpFrequency.Visible = False
```
if we set a control's visibility to false and then true, then will the control potentially flicker on the screen?
Would it be cleaner to use an approach similar to the approaches in my comments above?

Ivanluv replied:
I have corrected that

lloyddewit commented on instat/dlgDescribeTwoVariable.vb:
```diff
@@ -598,7 +953,256 @@ Public Class dlgDescribeTwoVariable
+
+    Private Sub ucrReceiverSecondTwoVariableFactor_SelectionChanged(sender As Object, e As EventArgs) Handles ucrReceiverSecondTwoVariableFactor.SelectionChanged
+        MatchingCategoricalByCategoricalVariables(ucrReceiverSecondTwoVariableFactor)
+    End Sub
+    Private Sub ucrReceiverFirstVars_SelectionChanged(sender As Object, e As EventArgs) Handles ucrReceiverFirstVars.SelectionChanged
+        MatchingCategoricalByCategoricalVariables(ucrReceiverFirstVars)
+    End Sub
+    Private Sub ucrReceiverThreeVariableSecondFactor_SelectionChanged(sender As Object, e As EventArgs) Handles ucrReceiverThreeVariableSecondFactor.SelectionChanged
```
empty lines between subs?

Ivanluv replied:
Thank you for this. Done

lloyddewit commented on instat/dlgDescribeTwoVariable.vb:
```diff
@@ -598,7 +953,256 @@ Public Class dlgDescribeTwoVariable
+            Exit Sub
+        End If
+        Dim bContainedInMultipleReceiver As Boolean = False
+        Dim strWarningMessage As String = ""
```
delete?

Ivanluv replied:
Its no longer being used after you suggestions below. Its deleted

lloyddewit commented on instat/dlgDescribeTwoVariable.vb:
```diff
@@ -598,7 +953,256 @@ Public Class dlgDescribeTwoVariable
+
+        If sender Is ucrReceiverSecondTwoVariableFactor Then
+            If bContainedInMultipleReceiver And strFirstVariablesType = "categorical" Then
+                strWarningMessage = " First Variable "
+                DisplayWarning(strWarningMessage)
```
pass literal string to `DisplayWarning()`? (also in 4 other places below)
```suggestion
                DisplayWarning(" First Variable ")
```

Ivanluv replied:
Done

lloyddewit commented on instat/dlgDescribeTwoVariable.vb:
```diff
@@ -598,7 +953,256 @@ Public Class dlgDescribeTwoVariable
+        ElseIf sender Is ucrReceiverThreeVariableThirdVariable Then
+            If (bContainedInMultipleReceiver And strFirstVariablesType = "categorical") OrElse
+               (ucrReceiverThreeVariableThirdVariable.GetVariableNames = ucrReceiverThreeVariableSecondFactor.GetVariableNames) Then
+                strWarningMessage = " First Variable and the Second Variable "
```
```suggestion
                strWarningMessage = " First Variable and Second Variable "
```

Ivanluv replied:
Done

lloyddewit commented on instat/dlgDescribeTwoVariable.vb:
```diff
@@ -598,7 +953,256 @@ Public Class dlgDescribeTwoVariable
+        ElseIf sender Is ucrReceiverThreeVariableSecondFactor Then
+            If (strFirstVariablesType = "categorical" AndAlso bContainedInMultipleReceiver) OrElse
+                              (ucrReceiverThreeVariableThirdVariable.GetVariableNames = ucrReceiverThreeVariableSecondFactor.GetVariableNames) Then
+                strWarningMessage = " First Variable and the third Variable "
```
```suggestion
                strWarningMessage = " First Variable and Third Variable "
```

Ivanluv replied:
Done

lloyddewit commented on instat/dlgDescribeTwoVariable.vb:
```diff
@@ -598,7 +953,256 @@ Public Class dlgDescribeTwoVariable
+                End If
+            ElseIf rdoThreeVariable.Checked Then
+                If bContainedInMultipleReceiver Then
+                    strWarningMessage = " Second and Third Variable "
```
```suggestion
                    strWarningMessage = " Second Variable and Third Variable "
```

Ivanluv replied:
Done

lloyddewit reviewed (changes requested):
@Ivanluv Thank you, this looks like a big piece of work.
My comments below

Ivanluv commented on instat/dlgDescribeTwoVariable.vb:
```diff
@@ -130,464 +181,768 @@ Public Class dlgDescribeTwoVariable
+        Else
+            Return False
+        End If
+    End Function
```
This is the best approach. Done

lloyddewit commented:
@rdstern if you can test/approve, then we can merge, thanks

lilyclements commented:
@rdstern can you test/approve this? I think it is ready to go in, and it would be great to have these changes in for the workshop next week

rdstern reviewed (commented):
@lilyclements it is looking close.  But the third receiver is still limited to being a factor.  That is not clear in the numerical/categorical by categorical, by ? summary on the left.  So, I thought I could get summary tables by putting the numeric variables in the multiple receiver - which I can.  Or I could have factors in the multiple receiver, and another factor in the second, and then a numeric variable in the third - which I can't.  Then I would get multiple summary tables of the particularly important numeric variable by various factors.  It looks as though that should be possible and I suggest it would be useful.

lilyclements commented:
@rdstern this has been sorted now - it works my end, so hopefully you agree!

rdstern commented:
a) When I pressed reset I got this message.  There seem to be various messages and it would be good to check if they are really needed.  This one was certainly not relevant here.
![image](https://user-images.githubusercontent.com/11226469/188842564-015d755a-4a66-4dcd-91bd-67f12f681e14.png)
b) The second receiver doesn't advance once filled:  (It does with the Skim option at the top, but not here:
![image](https://user-images.githubusercontent.com/11226469/188843109-b04145e5-57f5-4370-8c9c-ed1d70b39e5e.png)
c) The numerical summaries don't come!  Here is a result from Village by Variety by Yield
![image](https://user-images.githubusercontent.com/11226469/188843509-5e7025be-5d1c-4899-b3dd-8b45aa5a2f7b.png)

N-thony commented:
@Vitalis95 would you like to take this over? I'm happy to support you.

Vitalis95 commented:
@lilyclements , I am trying to look into this issue- for part c) above, Factor (many) by Factor, by Numeric ; 
 i.e `village` by `variety` by `yield ` here is the code;
```
# Code generated by the dialog, Describe Two Variables
survey <- data_book$get_data_frame(data_name="survey")
multiple_receiver_cols <- c("village")
list_of_tables  <- survey %>% purrr::map(.x=multiple_receiver_cols, .f= ~ dplyr::rename(.data=data_book$summary_table("survey", summaries=c("summary_mean", "summary_sd", "summary_sum"), columns_to_summarise=.x, factors=c("variety", .x), j=1), by_var=.x)) %>% purrr::map(.x=., .f= ~ mmtable2::mmtable(data=.x, cells=value) + mmtable2::header_top_left(variable='by_var') + mmtable2::header_left_top(variable="variety") + mmtable2::header_left_top(variable="summary-variable"))
```
it seems the function only finds summaries on the variable(s) in the multiple receiver(`multiple_receiver_cols `) by the variables in other receivers.
For this case, it is doing summaries for `village` (multiple_receiver_cols) which is a factor that is why mean, sd are empty
Also I found that the numeric variable `yield` added in the third receiver is not in the code
@lilyclements is there a way we can  get summary tables by having  factors in the multiple receiver, and another factor in the second, and then a numeric variable in the third receiver like the way we have numeric variables in  multiple receiver by factor in second and third receivers? thanks
